From 8dce37bb09e889e7827145029304025904bf6df1 Mon Sep 17 00:00:00 2001
From: Olaf Lessenich <lessenic@fim.uni-passau.de>
Date: Mon, 13 Jul 2015 15:10:53 +0200
Subject: [PATCH 18/18] Extend classes for matching.

Signed-off-by: Olaf Lessenich <lessenic@fim.uni-passau.de>
---
 java4/frontend/JDime.jadd | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/java4/frontend/JDime.jadd b/java4/frontend/JDime.jadd
index d9a46c4..d92498d 100644
--- a/java4/frontend/JDime.jadd
+++ b/java4/frontend/JDime.jadd
@@ -82,4 +82,20 @@ aspect JDime {
 		if (right != null) { right.flushCaches(); }
 	}
 
+	public String ASTNode.getMatchingRepresentation() {
+		return dumpString();
+	}
+
+	public String ImportDecl.getMatchingRepresentation() {
+		return prettyPrint();
+	}
+
+	public String Literal.getMatchingRepresentation() {
+		return prettyPrint();
+	}
+
+	public boolean ASTNode.matches(ASTNode other) {
+		return getClass().equals(other.getClass())
+		       && getMatchingRepresentation().equals(other.getMatchingRepresentation());
+	}
 }
-- 
2.1.4

